### PR TITLE
[15][MIG] partner_contact_in_several_companies

### PR DESCRIPTION
**Pull request by dessanhemrayev**
Migrate v14 to v15

### CODE REVIEW

File: partner_contact_in_several_companies/models/ir_actions.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+class IRActionsWindow(models.Model):
+    _inherit = "ir.actions.act_window"
+
+    def read(self, fields=None, context=None, load="_classic_read"):
+        actions = super(IRActionsWindow, self).read(fields=fields, load=load)
+        for action in actions:
+            if action.get("res_model", "") == "res.partner":
+                # By default, only show standalone contact
+                action_context = action.get("context", "{}")
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/ir_actions.py]: Need to convert context value to dict:
`safe_eval(action.get("context", "{}"))`

File: partner_contact_in_several_companies/models/ir_actions.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                if "search_show_all_positions" not in action_context:
+                    action["context"] = action_context.replace(
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/ir_actions.py]: Replacing string is not a good solution. Better to convert context to dict and update dict value.

`action["context"].update({...})`

File: partner_contact_in_several_companies/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        else:
+            result = self
+        return result
+
+    @api.model
+    def search(self, args, offset=0, limit=None, order=None, count=False):
+        """Display only standalone contact matching ``args`` or having
+        attached contact matching ``args``"""
+        ctx = self.env.context
+        if (
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py]: If you use construction `dict[key]` then you may get except KeyError. Better use `get()` function for getting dicts entry (`dict.get(key, default value)`).

Add the variable:
```
searsearch_show_all_positions = ctx.get("search_show_all_positions", {})
```
And then:
```
if searsearch_show_all_positions.get("is"set") and not searsearch_show_all_positions.get("set_value"):
```

File: partner_contact_in_several_companies/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+    @api.model
+    def create(self, vals):
+        """When creating, use a modified self to alter the context (see
+        comment in _basecontact_check_context).  Also, we need to ensure
+        that the name on an attached contact is the same as the name on the
+        contact it is attached to."""
+        modified_self = self._basecontact_check_context("create")
+        if not vals.get("name") and vals.get("contact_id"):
+            vals["name"] = modified_self.browse(vals["contact_id"]).name
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py]: Better use get() function for dict
```
contact_id = vals.get("contaxt_id")
if not vals.get("name") and contact_id:
    vals.update(name=modified_self.browse(contact_id)).name)
```

File: partner_contact_in_several_companies/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    def unlink(self):
+        modified_self = self._basecontact_check_context("unlink")
+        return super(ResPartner, modified_self).unlink()
+
+    def _compute_commercial_partner(self):
+        """Returns the partner that is considered the commercial
+        entity of this partner. The commercial entity holds the master data
+        for all commercial fields (see :py:meth:`~_commercial_fields`)"""
+        result = super(ResPartner, self)._compute_commercial_partner()
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py]: The computed method doesn't return a value
```
super(ResPartner, self)._compute_commercial_partner()
```

File: partner_contact_in_several_companies/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        for partner in self:
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py:102]: In cycle will may be apply filter function for removing nesting.
```
for partner in self.filtered(lambda p: p.contact_type == "attached" and not p.parent_id):
```

File: partner_contact_in_several_companies/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            if partner.contact_type == "attached" and not partner.parent_id:
+                partner.commercial_partner_id = partner.contact_id
+        return result
+
+    def _contact_fields(self):
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py:107]: This function don't use res.partner entries that is why add  `@api.model` decorator.

File: partner_contact_in_several_companies/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        that the name on an attached contact is the same as the name on the
+        contact it is attached to."""
+        modified_self = self._basecontact_check_context("create")
+        if not vals.get("name") and vals.get("contact_id"):
+            vals["name"] = modified_self.browse(vals["contact_id"]).name
+        return super(ResPartner, modified_self).create(vals)
+
+    def read(self, fields=None, load="_classic_read"):
+        modified_self = self._basecontact_check_context("read")
+        return super(ResPartner, modified_self).read(fields=fields, load=load)
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py:87]: For what create variable `modified_self`?  May be insert in `super()`?

```super(ResPartner, self._basecontact_check_context("read")).read(fields=fields, load=load)```

File: partner_contact_in_several_companies/models/res_partner.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+    def write(self, vals):
+        modified_self = self._basecontact_check_context("write")
+        return super(ResPartner, modified_self).write(vals)
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py:91]: `super(ResPartner, self._basecontact_check_context("write")).write(vals)`

File: partner_contact_in_several_companies/models/res_partner.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+    def unlink(self):
+        modified_self = self._basecontact_check_context("unlink")
+        return super(ResPartner, modified_self).unlink()
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py:95]: `super(ResPartner, self._basecontact_check_context("unlink")).unlink()`

**Review by geomer198 — CHANGES_REQUESTED** [review]
Refactor the module code.

File: partner_contact_in_several_companies/models/ir_actions.py (hunk 1 above)
Comment (dessanhemrayev) [reply partner_contact_in_several_companies/models/ir_actions.py]: Got it, I'll fix it

File: partner_contact_in_several_companies/models/ir_actions.py (hunk 2 above)
Comment (dessanhemrayev) [reply partner_contact_in_several_companies/models/ir_actions.py]: Got it, I'll fix it

File: partner_contact_in_several_companies/models/res_partner.py (hunk 3 above)
Comment (dessanhemrayev) [reply partner_contact_in_several_companies/models/res_partner.py]: Thanks for the example, I will fix it

File: partner_contact_in_several_companies/models/res_partner.py (hunk 4 above)
Comment (dessanhemrayev) [reply partner_contact_in_several_companies/models/res_partner.py]: Thanks for the example, I will fix it

File: partner_contact_in_several_companies/models/res_partner.py (hunk 5 above)
Comment (dessanhemrayev) [reply partner_contact_in_several_companies/models/res_partner.py]: Got it, I'll fix it

File: partner_contact_in_several_companies/models/res_partner.py (hunk 7 above)
Comment (dessanhemrayev) [reply partner_contact_in_several_companies/models/res_partner.py:107]: Got it, I'll fix it

File: partner_contact_in_several_companies/models/res_partner.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+    def _compute_commercial_partner(self):
+        """Returns the partner that is considered the commercial
+        entity of this partner. The commercial entity holds the master data
+        for all commercial fields (see :py:meth:`~_commercial_fields`)"""
+        super(ResPartner, self)._compute_commercial_partner()
+        for partner in self:
+            if partner.contact_type == "attached" and not partner.parent_id:
+                partner.commercial_partner_id = partner.contact_id
+        return True
```
Comment (geomer198) [inline partner_contact_in_several_companies/models/res_partner.py]: Computing function return None that is why can you delete this return?

**Review by geomer198 — APPROVED** [review]

**Comment by Aldeigja** [comment]
@OCA/crm-sales-marketing-maintainers could you review?

**Comment by Aldeigja** [comment]
https://github.com/orgs/OCA/teams/crm-sales-marketing-maintainers gentle reminder :)

**Comment by Aldeigja** [comment]
@OCA/crm-sales-marketing-maintainers please review when you have a moment. Thank you!

**Review by aayartsev — APPROVED** [review]
For me all is good

**Comment by Aldeigja** [comment]
@OCA/crm-sales-marketing-maintainers is good for merge?

**Comment by mmequignon** [comment]
@dessanhemrayev hello!
I think smth wrong happened during the migration.
The git history has been lost, see this https://github.com/OCA/partner-contact/commits/14.0/partner_contact_in_several_companies
Could you please follow this guide https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0 ?

Thank you!

**Comment by stephansainleger** [comment]
@aayartsev @mmequignon I need this addon to be migrated in 15.0. As this PR seems to be blocked, I have created a new one with commits squashed: https://github.com/OCA/partner-contact/pull/1702
If it is ok with OCA policy (I'm newby as contributor), could you please validate it? Thanks :-)

**Pull request closed without merging**